docs/shared/guides/use-environment-variables-in-angular.md

```diff
@@ -1,10 +1,187 @@
+import { Injectable } from '@angular/core';
+
+@Injectable({ providedIn: 'root' })
+export class ApiService {
+  constructor() {
+    console.log('API URL:', MY_API_URL);
+  }
+}
+```
+
+You can also define the variables in a way that allows to consume them as you would do in Node.js applications:
+
+```json {% fileName="apps/my-app/project.json" highlightLines=[5,"8-10"] %}
+{
+  ...
+  "targets": {
+    "build": {
+      "executor": "@nx/angular:application",
+      "options": {
+        ...
+        "define": {
+          "process.env.MY_API_URL": "http://localhost:3333"
+        }
+      }
+    },
+    ...
+  }
+}
+```
+
+Like the previous example, you must configure TypeScript to recognize the `process.env` object. You can do this by defining the `process.env` object in a type definition file:
+
+```ts {% fileName="apps/my-app/src/types.d.ts" %}
+declare const process: {
+  env: {
+    API_URL: string;
+  };
+};
+```
+
+{% callout type="warning" title="Beware" %}
+You could also add the Node.js types to your `tsconfig.json` file, but this would add many types that you don't need in a browser environment. This can be misleading since you'll see some types are available, but the runtime functionality won't be. It's better to define only the types you need in a type definition file.
+{% /callout %}
+
+And then use the variable in your application code:
+
+```ts {% fileName="apps/my-app/src/app/api.service.ts" highlightLines=[6] %}
+import { Injectable } from '@angular/core';
+
+@Injectable({ providedIn: 'root' })
+export class ApiService {
+  constructor() {
+    console.log('API URL:', process.env.MY_API_URL);
+  }
+}
+```
+
+### Using a custom ESBuild plugin
+
+{% callout type="note" title="Required Angular version" %}
+Support for custom ESBuild plugins requires Angular **17.0.0** or later.
+{% /callout %}
+
+The previous method is useful to statically define variables in the project configuration that will be available at build time. However, if you need to dynamically collect and define the environment variables available at build time, you can create a custom ESBuild plugin.
+
+You can provide custom ESBuild plugin to the `@nx/angular:application` or `@nx/angular:browser-esbuild` executors:
+
+```json {% fileName="apps/my-app/project.json" highlightLines=[5,8] %}
+{
+  ...
+  "targets": {
+    "build": {
+      "executor": "@nx/angular:application",
+      "options": {
+        ...
+        "plugins": ["apps/my-app/plugins/env-var-plugin.js"]
+      }
+    },
+    ...
+  }
+}
+```
+
+Next, create the custom ESBuild plugin:
+
+```js {% fileName="apps/my-app/plugins/env-var-plugin.js" %}
+const myOrgEnvRegex = /^MY_ORG_/i;
+
+const envVarPlugin = {
+  name: 'env-var-plugin',
+  setup(build) {
+    const options = build.initialOptions;
+
+    const envVars = {};
+    for (const key in process.env) {
+      if (myOrgEnvRegex.test(key)) {
+        envVars[key] = JSON.stringify(process.env[key]);
```

Comment:
sorry for jumping in, but isn't `JSON.stringify(process.env[key])` superfluous in this case, since all environment variables from process.env are strings anyway?

Reply:
It's indeed superfluous, but not because they are string values. A string value and a JSON stringified string value are not the same thing. Still, we don't need to stringify each value because we already stringify the full `envVars` object when assigning it.